### ISSUE TITLE
Previewer function is not notified of all keystroke events

### ISSUE THREAD

**Issue by AmanNegi** [RocketChat/Rocket.Chat#32650]
### Description:

I am implementing the `previewer` function of the slash command. While reading the arguments available in the function, I noticed that the arguments are not updated/notified beyond a certain point. 

### Steps to reproduce:

1. Generate a RC-App and create a class that implements the `ISlashCommand`.
2. Implement the `previewer` function, and simply get the arguments using `context.getArguments()`. 
3. Print the arguments to the console (or logs). You will be able to see the arguments instantaneously in the terminal thus I prefer to print to console. Print to console, only if you have a server running on your Local System/Docker.
4. Upload the app to the RC server, and test your `/command args`. 

### Expected behavior:

The `previewer` function should be notified of all changes made to the text field, so that operations can be performed on the latest and intermediate changes.

### Actual behavior:

Currently, as mentioned above the `previewer` function is not notified of all updates made to the text field, even when the user stops writing it doesn't get updated. Which is unexpected and reduces the possibilities of utilizing the `previewer` function.

### Video Demonstration:

https://github.com/RocketChat/Rocket.Chat/assets/37607224/6e4312fe-8c9f-4f4d-8a4f-5f68c3a3f5a8

### Server Setup Information:

- Version of Rocket.Chat Server: 6.9.0
- Operating System: Linux
- Deployment Method: Docker

### Client Setup Information

- Desktop App or Browser Version: Firefox 127.0
- Operating System: Ubuntu 22.04

### Additional context

When I incurred this issue, I looked into other RC apps which utilized the `previewer` function and one of them was [Apps.Giphy](https://github.com/wreiske/Rocket.Chat.App-Giphy) however upon testing it locally I was able to conclude it also faces similar issues.

**Comment by reetp** [comment]
Hi.

Thanks for reporting.

Can you test on the latest dev code please?

Thanks.

**Comment by AmanNegi** [comment]
Hey @reetp I tried it out on the develop branch `6.10.0-develop` and it does face the same issue.

**Comment by reetp** [comment]
Ok thanks for testing.

I'll make sure someone is aware.

This may need moving to the apps engine repo but I'll check.

**Comment by reetp** [comment]
Ok someone will take a look..

No promises on a fix yet :-)

**Comment by AmanNegi** [comment]
I was able to diagnose this issue and was able to find a few pointers and fixes.

## Error & Logs :computer: :warning: 

#### Snapshot of browser console while typing `/prt a long command here`
![Screenshot from 2024-07-31 11-00-43](https://github.com/user-attachments/assets/f5060d2d-9d59-4d1c-ab84-7c027c7f1986)
> TLDR: `error: "too-many-requests", reason: "Error, too many requests. Please slow down. You must wait 8 seconds before trying again.", message: "Error, too many requests. Please slow down. You must wait 8 seconds before trying again. [too-many-requests]"` 

## My Opinion :thinking: :rocket:  
A simple fix to this is using debouncing, which would significantly reduce the number of API Calls and remove this error. If I were to pinpoint the line causing this issue it would be [this](https://github.com/RocketChat/Rocket.Chat/blob/71af51487365726900914d9776d0764463a91b20/apps/meteor/client/views/room/providers/ComposerPopupProvider.tsx#L357C1-L357C69). The debouncing approach is commonly used in text-editors which have to deal with users entering text too frequently and reducing the back-end load.

## Fix using Debouncing :wrench: :rocket: 

[![YouTube](http://i.ytimg.com/vi/UDa63z8Wz_Y/hqdefault.jpg)](https://www.youtube.com/watch?v=UDa63z8Wz_Y)

### Approach:

So I identified that the issue lies in continuous requesting the back-end, thus I created a de-bounced implementation of the `call()` function. Which awaits for 2 seconds(currently) and if by the end of 2 seconds there are no requests it continues with the request to update the backend about the changes. However, if there is any new request before 2 seconds end, the last request is cancelled. This fixes the issue and the error logs in browser also dissapears.

@reetp  Could you please look into this and suggest if I should create a PR for this with any changes you would prefer.

**Comment by reetp** [comment]
You don't need to @ anyone thanks.

We read the issues.

Appreciate your efforts.

I have referred this to the dev team. They are the ones who decide.

Rocket.Chat is a large organisation and it takes time to schedule things. 

Not much else I can say except be patient.

**Comment by AmanNegi** [comment]
Okay, I'll ensure that and thanks for replying. ✨🙌

**Comment by reetp** [comment]
> Okay, I'll ensure that and thanks for replying. ✨🙌

NP. The least I can do :-)

It is on our watch list.

I have raised this again with the dev team and I hope they'll get back to you soon.
